packages/data-service/src/data-service.ts

```diff
@@ -838,27 +836,175 @@ class DataServiceImpl extends WithLogContext implements DataService {
+
+      for (const fieldName of Object.keys(shardStorageStats)) {
+        if (
+          ['ns', 'ok', 'lastExtentSize', 'paddingFactor'].includes(fieldName)
+        ) {
+          continue;
+        }
+        if (
+          [
+            'userFlags',
+            'capped',
+            'max',
+            'paddingFactorNote',
+            'indexDetails',
+            'wiredTiger',
+          ].includes(fieldName)
+        ) {
+          // Fields that are copied from the first shard only, because they need to
+          // match across shards.
+          result[fieldName] ??= shardStorageStats[fieldName];
+        } else if (
+          // NOTE: `numOrphanDocs` is new in 6.0. `totalSize` is new in 4.4.
+          [
+            'count',
+            'size',
+            'storageSize',
+            'totalIndexSize',
+            'totalSize',
+            'numOrphanDocs',
+          ].includes(fieldName)
+        ) {
+          if (counts[fieldName] === undefined) {
+            counts[fieldName] = 0;
+          }
+          counts[fieldName] += coerceToJSNumber(shardStorageStats[fieldName]);
+        } else if (fieldName === 'avgObjSize') {
+          const shardAvgObjSize = coerceToJSNumber(
+            shardStorageStats[fieldName]
+          );
+          unscaledCollSize += shardAvgObjSize * shardObjCount;
+        } else if (fieldName === 'maxSize') {
+          const shardMaxSize = coerceToJSNumber(shardStorageStats[fieldName]);
+          maxSize = Math.max(maxSize, shardMaxSize);
+        } else if (fieldName === 'indexSizes') {
+          for (const indexName of Object.keys(shardStorageStats[fieldName])) {
+            if (indexSizes[indexName] === undefined) {
+              indexSizes[indexName] = 0;
+            }
+            indexSizes[indexName] += coerceToJSNumber(
+              shardStorageStats[fieldName][indexName]
+            );
+          }
+        } else if (fieldName === 'nindexes') {
+          const shardIndexes = shardStorageStats[fieldName];
+
+          if (nindexes === 0) {
+            nindexes = shardIndexes;
+          } else if (shardIndexes > nindexes) {
+            // This hopefully means we're building an index.
+            nindexes = shardIndexes;
+          }
+        }
+      }
+    }
+
+    if (collStats[0].shard) {
+      result.shards = shardStats;
+    }
+
+    try {
+      result.sharded = !!(await this._collection(
+        'config.collections',
+        'CRUD'
+      ).findOne({
+        _id: ns as any,
+        // Dropped is gone on newer server versions, so check for !== true
+        // rather than for === false (SERVER-51880 and related).
+        dropped: { $ne: true },
+      }));
+    } catch (e) {
+      // A user might not have permissions to check the config. In which
+      // case we default to the potentially inaccurate check for multiple
+      // shard response documents to determine if the collection is sharded.
+      result.sharded = collStats.length > 1;
+    }
```

Comment:
I don’t think we need this part here?

I think generally speaking it would be nice if we could reduce this method to support only what’s used by `_buildCollectionStats` and then see if maybe we can push all or most of what remains into the aggregation itself, rather than aggregating client-side

Reply:
I reviewed properties and refactored this method, but I don't feel confident enough to rewrite it with a poor aggregation 😀🙈

Reply:
Fwiw, there’s a bunch more fields that look like they can be removed: e0b6eadd8ed31644b1ec11e6e77a78c63d02d416

And I did try to spell out how to do this in a pipeline rather than client-side code. I think this works but it might be good to check that against a real sharded cluster, if we want to try it :) c6501df6ec61cfee29c353cc6221637e27ce486d

Reply:
I merged your aggregation and tested on a local sharded cluster, seems fine 🎊